### ISSUE TITLE
Handling Axios timeouts (offline or API down) and notify user

### ISSUE THREAD

**Issue by jeremyhalin** [thecodingmachine/react-native-boilerplate#164]
I would like to be able to handle timeout API calls with Axios in `Services\index.js`.

I'm able to know when I have a timeout, but I don't know how to handle the error to display a message on the user's UI.
I know I can use `handleError` to save my error in state, but I need to do this for each action.
How can we find something global?

**Comment by JeremyDolle** [comment]
In the boilerplate we don't provide an UI kit for displaying errors. This is because we don't want to force developers using a library. On some of our projects we use [this](https://github.com/testshallpass/react-native-dropdownalert) library, you can use it like a service and use it in axios interceptor for example.

**Comment by jeremyhalin** [comment]
Well sorry for misunderstanding, I was not talking about UI kit for displaying errors.
My problem is relative to how we can trigger that kind of library (or any code btw), when something is catched in axios interceptor.
In `Services\index.js` I don't know how to trigger something to display a message on UI or to update state to be able to listen to a `api.error` state for example on a top level navigator like `Application.js`.

In short: I want to display an error message anytime axios interceptors "detect" API errors (especially 500 or timeouts).

> you can use it like a service

What do you mean? With `ref`? Like `navigationRef` in `Navigators\Root.js`? 😮

**Comment by jeremyhalin** [comment]
Alright I managed to display a flash message with [react-native-flash-message](https://github.com/lucasferreira/react-native-flash-message) library. Now when I my API isn't accessible or if user's phone is offline, I can display an error message about it.

I also used `i18next` and `react-i18next` already in this boilerplate to get a translated string with :
```js
import i18n from '@/Translations/index'

i18n.t('errors.unknown')
```
Just be sure to have `i18n` initialized before.

`src\App.js`
```js
import 'react-native-gesture-handler'
import React from 'react'
import { Provider } from 'react-redux'
import { PersistGate } from 'redux-persist/lib/integration/react'
import { store, persistor } from '@/Store'
import { ApplicationNavigator } from '@/Navigators'
import './Translations'
import FlashMessage from 'react-native-flash-message'

const App = () => (
  <Provider store={store}>
    {/**
     * PersistGate delays the rendering of the app's UI until the persisted state has been retrieved
     * and saved to redux.
     * The `loading` prop can be `null` or any react instance to show during loading (e.g. a splash screen),
     * for example `loading={<SplashScreen />}`.
     * @see https://github.com/rt2zz/redux-persist/blob/master/docs/PersistGate.md
     */}
    <PersistGate loading={null} persistor={persistor}>
      <ApplicationNavigator />
      <FlashMessage position="bottom" floating={true} icon={'auto'} />
    </PersistGate>
  </Provider>
)

export default App
```

`src\Services\index.js`
```js
import axios from 'axios'
import { Config } from '@/Config'
import * as Keychain from 'react-native-keychain'
import { navigate } from '@/Navigators/Root'
import { showMessage } from 'react-native-flash-message'
import i18n from '@/Translations/index'

const instance = axios.create({
  baseURL: Config.API_URL,
  headers: {
    Accept: 'application/json',
    'Content-Type': 'application/json',
  },
  timeout: 3000,
})

export const handleError = ({ message, data, status }) => {
  return Promise.reject({ message, data, status })
}

instance.interceptors.request.use(
  async function (config) {
    // Do something before request is sent
    const accessToken = await Keychain.getInternetCredentials('access_token')
    config.headers.common.Authorization = `Bearer ${accessToken.password}`
    return config
  },
  function (error) {
    console.log('error request', error)
    // Do something with request error
    return Promise.reject(error)
  },
)

// Add a response interceptor
instance.interceptors.response.use(
  function (response) {
    // Any status code that lie within the range of 2xx cause this function to trigger
    // Do something with response data
    return response
  },
  async function (error) {
    if (error.response) {
      const originalRequest = error.config
      console.log('originalRequest._retry=', originalRequest._retry)
      // The request was made and the server responded with a 403 status code
      // which means access_token is expired, so we try to get a new access_token
      // from the refresh_token and retry request
      if (error.response.status === 403 && !originalRequest._retry) {
        originalRequest._retry = true

        const refreshToken = await Keychain.getInternetCredentials(
          'refresh_token',
        )

        const response = await instance.post('auth/refresh', {
          refresh_token: refreshToken.password,
        })

        await Keychain.setInternetCredentials(
          'access_token',
          'access_token',
          response.data.data.attributes.token,
        )

        await Keychain.setInternetCredentials(
          'refresh_token',
          'refresh_token',
          response.data.included.attributes.token,
        )

        const accessToken = {
          token: response.data.data.attributes.token,
          expiration: response.data.data.attributes.expiration,
        }

        axios.defaults.headers.common.Authorization = `Bearer ${accessToken.token}`
        return instance(originalRequest)
      } else if (originalRequest.url === 'auth/refresh') {
        // error while trying to refresh access token, so disconnect!!
        await Keychain.resetInternetCredentials('access_token')
        await Keychain.resetInternetCredentials('refresh_token')
        navigate('Auth', { screen: 'IndexAuth' })
      }

      let errorMessage = 'Response status code <> 200 (' + error.message + ')'
      if (
        error.response.data.errors &&
        error.response.data.errors[0] &&
        error.response.data.errors[0].title
      ) {
        errorMessage = error.response.data.errors[0].title
      }

      // The request was made and the server responded with a status code
      // that falls out of the range of 2xx
      return handleError({
        message: errorMessage,
        status: error.response.status,
        data: error.response.data,
      })
    } else if (error.request) {
      // The request was made but no response was received
      // `error.request` is an instance of XMLHttpRequest in the browser and an instance of
      // http.ClientRequest in node.js

      showMessage({
        message: i18n.t('errors.offline'),
        type: 'danger',
        duration: 5000,
      })

      return handleError({
        message: 'No response received (' + error.message + ')',
      })
    } else {
      // Something happened in setting up the request that triggered an Error

      showMessage({
        message: i18n.t('errors.unknown'),
        type: 'danger',
        duration: 5000,
      })

      return handleError({
        message: 'Unknown error (' + error.message + ')',
      })
    }
  },
)

export default instance
```
